kotlinx-coroutines-core/jvm/src/CoroutineContext.kt

```diff
@@ -10,35 +10,73 @@ import kotlin.coroutines.jvm.internal.CoroutineStackFrame
 
 /**
  * Creates context for the new coroutine. It installs [Dispatchers.Default] when no other dispatcher nor
- * [ContinuationInterceptor] is specified, and adds optional support for debugging facilities (when turned on).
+ * [ContinuationInterceptor] is specified, and adds optional support for
+ * copyable thread context [elements][CopyableThreadContextElement] and  debugging facilities (when turned on).
  *
  * See [DEBUG_PROPERTY_NAME] for description of debugging facilities on JVM.
  */
 @ExperimentalCoroutinesApi
 public actual fun CoroutineScope.newCoroutineContext(context: CoroutineContext): CoroutineContext {
-    val combined = coroutineContext.foldCopiesForChildCoroutine() + context
+    val combined = foldCopies(coroutineContext, context, true)
     val debug = if (DEBUG) combined + CoroutineId(COROUTINE_ID.incrementAndGet()) else combined
     return if (combined !== Dispatchers.Default && combined[ContinuationInterceptor] == null)
         debug + Dispatchers.Default else debug
 }
 
 /**
- * Returns the [CoroutineContext] for a child coroutine to inherit.
- *
- * If any [CopyableThreadContextElement] is in the [this], calls
- * [CopyableThreadContextElement.copyForChildCoroutine] on each, returning a new [CoroutineContext]
- * by folding the returned copied elements into [this].
- *
- * Returns [this] if `this` has zero [CopyableThreadContextElement] in it.
+ * Creates context for coroutine builder functions that do not launch a new coroutine,
+ * but change current coroutine context, such as [withContext] and `runBlocking`.
  */
-private fun CoroutineContext.foldCopiesForChildCoroutine(): CoroutineContext {
-    val hasToCopy = fold(false) { result, it ->
-        result || it is CopyableThreadContextElement<*>
+@ExperimentalCoroutinesApi
+public actual fun CoroutineContext.newCoroutineContext(addedContext: CoroutineContext): CoroutineContext {
+    return foldCopies(this, addedContext, false)
+}
+
+private val hasCopyableElements: (Boolean, CoroutineContext.Element) -> Boolean = { result, it ->
+    result || it is CopyableThreadContextElement<*>
+}
+
+/*
+ * Folds two contexts if there is need to.
+ */
+private fun foldCopies(originalContext: CoroutineContext, appendContext: CoroutineContext, isNewCoroutine: Boolean): CoroutineContext {
```

Comment:
Effectively, we are trying to provide here an extension point to implementing a mutable (sic) `CoroutineLocal` without actually giving a concrete implementation of it. The reason for not giving a concrete implementation, as I see it, is that there seem to be two different use-cases for mutable coroutine locals:
* Like in your case, you may have an existing system that keeps some "source of truth" in some behind-the-scenes thread-local storage and all you need to do is do synchronize with it. If we just gave you a mutable `CoroutineLocal`, then it would not have helped in writing this synchronization logic.
* We might envision some genuine need to have mutable coroutine locals that are not synchronized to any thread-local storage. 

An interesting observation here is that if we just limit ourselves to those two use-cases, then we may not need two separate `copy` and `merge` methods at the end.

Reply:
> We might envision some genuine need to have mutable coroutine locals that are not synchronized to any thread-local storage.

An unsynchronized mutable coroutine local might be sufficient in all cases.

A (classic, not `Copyable`) `ThreadContextElement` could read and write an adjacent thread-unsafe `CoroutineLocal`. With some "clever" typing, a single public `CoroutineContext.Element` type can probably be implemented by a pair of elements?

> An interesting observation here is that if we just limit ourselves to those two use-cases, then we may not need two separate copy and merge methods at the end.

I don't think separate methods are necessary.

> Effectively, we are trying to provide here an extension point to implementing a mutable (sic) CoroutineLocal without actually giving a concrete implementation of it.

It may be safer to give a concrete implementation than a way to supply an implementation. If an extension point is indeed "too grotesque to be actually useful", I might be cautious about entrusting my program's memory consistency to it.